cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -113,6 +132,66 @@ inline int64_t ValuesToBitmap(PyArrayObject* arr, uint8_t* bitmap) {
   return null_count;
 }
 
+class NumPyNullsConverter {
+ public:
+  /// Convert the given array's null values to a null bitmap.
+  /// The null bitmap is only allocated if null values are ever possible.
+  static Status Convert(MemoryPool* pool, PyArrayObject* arr,
+                        bool use_pandas_null_sentinels,
+                        std::shared_ptr<ResizableBuffer>* out_null_bitmap_,
+                        int64_t* out_null_count) {
+    NumPyNullsConverter converter(pool, arr, use_pandas_null_sentinels);
+    RETURN_NOT_OK(VisitNumpyArrayInline(arr, &converter));
+    *out_null_bitmap_ = converter.null_bitmap_;
+    *out_null_count = converter.null_count_;
+    return Status::OK();
+  }
+
+  template <int TYPE>
+  Status Visit(PyArrayObject* arr) {
+    typedef internal::npy_traits<TYPE> traits;
+
+    const bool null_sentinels_possible =
+        // Observing pandas's null sentinels
+        (use_pandas_null_sentinels_ && traits::supports_nulls);
+
+    if (null_sentinels_possible) {
+      RETURN_NOT_OK(InitNullBitmap(PyArray_SIZE(arr)));
+      null_count_ = ValuesToBitmap<TYPE>(arr, null_bitmap_data_);
+    }
+    return Status::OK();
+  }
+
+  // XXX it's the same as NumPyConverter::InitNullBitmap()
+  Status InitNullBitmap(int64_t length) {
+    int64_t null_bytes = BitUtil::BytesForBits(length);
+
+    null_bitmap_ = std::make_shared<PoolBuffer>(pool_);
+    RETURN_NOT_OK(null_bitmap_->Resize(null_bytes));
+
+    null_bitmap_data_ = null_bitmap_->mutable_data();
+    memset(null_bitmap_data_, 0, static_cast<size_t>(null_bytes));
```

Comment:
`std::fill(null_bitmap_data_, null_bitmap_data_ + null_bytes, 0)` is a bit more idiomatic.

Reply:
Hmm, perhaps. This is really a copy/paste of `NumPyConverter::InitNullBitmap()`...

Reply:
Possibly time for a subclass then?